src/mbgl/style/sources/geojson_source_impl.cpp

```diff
@@ -67,23 +67,55 @@ class SuperclusterData : public GeoJSONData {
     mapbox::supercluster::Supercluster impl;
 };
 
-GeoJSONSource::Impl::Impl(std::string id_, GeoJSONOptions options_)
-    : Source::Impl(SourceType::GeoJSON, std::move(id_)),
-      options(std::move(options_)) {
+template <class T>
+T EvaluateFeature(optional<T> accumulated,
+                  mapbox::feature::feature<double>& f,
+                  const std::shared_ptr<expression::Expression> expression,
+                  T finalDefaultValue = T()) {
+    const expression::EvaluationResult result = expression->evaluate(accumulated, f);
+    if (result) {
+        const optional<T> typed = expression::fromExpressionValue<T>(*result);
+        return typed ? *typed : finalDefaultValue;
+    }
+    return finalDefaultValue;
+}
+
+GeoJSONSource::Impl::Impl(std::string id_, optional<GeoJSONOptions> options_)
+    : Source::Impl(SourceType::GeoJSON, std::move(id_)) {
+        options = options_ ? std::move(*options_) : GeoJSONOptions{};
 }
 
-GeoJSONSource::Impl::Impl(const Impl& other, const GeoJSON& geoJSON)
-    : Source::Impl(other),
-      options(other.options) {
+GeoJSONSource::Impl::Impl(const Impl& other, const GeoJSON& geoJSON) : Source::Impl(other) {
+    options = std::move(other.options);
     constexpr double scale = util::EXTENT / util::tileSize;
-
-    if (options.cluster
-        && geoJSON.is<mapbox::feature::feature_collection<double>>()
-        && !geoJSON.get<mapbox::feature::feature_collection<double>>().empty()) {
+    if (options.cluster && geoJSON.is<mapbox::feature::feature_collection<double>>() &&
+        !geoJSON.get<mapbox::feature::feature_collection<double>>().empty()) {
         mapbox::supercluster::Options clusterOptions;
         clusterOptions.maxZoom = options.clusterMaxZoom;
         clusterOptions.extent = util::EXTENT;
         clusterOptions.radius = ::round(scale * options.clusterRadius);
+        clusterOptions.map =
+            [&](const mapbox::feature::property_map& properties) -> mapbox::feature::property_map {
+            mapbox::feature::property_map ret{};
+            for (const auto& p : options.clusterProperties) {
+                auto feature = mapbox::feature::feature<double>();
+                feature.properties = properties;
```

Comment:
could we introduce `evaluateFeature`, accepting properties, so that we don't have to copy `properties` on each iteration?

Reply:
The purpose of copying property to feature is that the `EvaluationContext` only supports feature with base type `GeometryTileFeature`. With our use case, we take use of `GeoJSONFeature`, I was also considering parsing directly with the properties, however, `GeoJSONFeature` has its `Feature` member as a const type. So either we change the constness or we keep the copying mechanism.